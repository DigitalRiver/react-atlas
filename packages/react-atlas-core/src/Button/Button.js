import React from "react";
import PropTypes from "prop-types";
import {IconCore} from './../Icon';
import cx from "classnames";

/**
 * A Generic button component.
 */
class Button extends React.PureComponent {
  constructor(props) {
    super(props);
  }

  render() {
    const {
      onClick,
      className,
      children,
      disabled,
      large,
      small,
      primary,
      secondary,
      warning,
      error,
      link,
      outline,
      icon,
      type,
      href,
      style
    } = this.props;

    let mainStyle = "button";
    if (primary) {
      mainStyle = "primary";
    } else if (secondary) {
      mainStyle = "secondary";
    } else if (warning) {
      mainStyle = "warning";
    } else if (error) {
      mainStyle = "error";
    } else if (link) {
      mainStyle = "link";
    } else {
      mainStyle = "default_btn";
    }

    if (outline) {
      mainStyle += "_outline";
    }

    /**
     * Use cx to construct a CSS class object.
     * cx will automatically remove falsy values
     * for us. This way only passed props will be
     * transformed into CSS classes.
     */
    const classes = cx(
      {
        "disabled": disabled,
        "large": large,
        "small": small,
        "button": true
      },
      mainStyle
    );

    let text = children;
    let iconClass = false;

    /* Just use the icon instead of text when an icon
     * is passed and no children(text) is set. */
    if (icon && children === "Default Button") {
      text = null;
    } else if (icon) {
      text = children;
      iconClass = "ra_Button__icon-left";
    }

    let btn = <button
          onClick={onClick}
          className={cx(className)}
          styleName={classes}
          style={style}
          type={type}
          href={href}
        >
          {icon ? <IconCore icon={icon} /> : null}
          {text}
        </button>
    ;

    let renderButton = btn;
    if (href) {
      renderButton = <a href={href}>{btn}</a>;
    }

    return renderButton;
  }
}

Button.propTypes = {
  /** HTML5 button type, eg submit, button, etc */
  "type": PropTypes.string,
  /** The URL to link to. */
  "href": PropTypes.string,
  /**
   * Define a mini button.
   *
   */
  "children": PropTypes.node,
  /**
   * Click event handler.
   */
  "onClick": PropTypes.func,
  /** An Object, array, or string of CSS classes to apply to button.*/
  "className": PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.array
  ]),
  /**
   *Determines if the button is disabled.
   * @examples <Button disabled/>
   */
  "disabled": PropTypes.bool,
  /**
   * use outline styled button
   * @examples <Button outline>
   */
  "outline": PropTypes.bool,
  /**
   * use primary style button
   * @examples <Button primary>
   */
  "primary": PropTypes.bool,
  /**
   * use secondary style button
   * @examples <Button secondary>
   */
  "secondary": PropTypes.bool,
  /**
   * use warning style button
   * @examples <Button warning>
   */
  "warning": PropTypes.bool,
  /**
   * use error style button
   * @examples <Button error>
   */
  "error": PropTypes.bool,
  /**
   * use link style button
   * @examples <Button link>
   */
  "link": PropTypes.bool,
  /**
   * use large style button
   * @examples <Button large>
   */
  "large": PropTypes.bool,
  /**
   * use small style button
   * @examples <Button small>
   */
  "small": PropTypes.bool,

  /**
   * The class name of the icon you want to set.
   */
  "icon": PropTypes.string,

  /**
   * Pass inline styles here.
   */
  "style": PropTypes.object
};

Button.defaultProps = {
  "children": "Default Button",
  "outline": false,
  "primary": false,
  "secondary": false,
  "warning": false,
  "error": false,
  "link": false,
  "large": false,
  "small": false,
  "disabled": false,
  "type": "button"
};

export default Button;
