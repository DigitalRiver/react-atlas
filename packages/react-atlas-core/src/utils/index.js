import events from "./events";
import prefixer from "./prefixer";
// import time from './time';
import utils from "./utils";
import classNames from "./classNames";

export { events, prefixer, utils, classNames };
