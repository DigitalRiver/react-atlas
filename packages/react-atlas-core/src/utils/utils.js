export default {
  angleFromPositions(cx, cy, ex, ey) {
    const theta = Math.atan2(ey - cy, ex - cx) + Math.PI / 2;
    return theta * 180 / Math.PI;
  },
  angle360FromPositions(cx, cy, ex, ey) {
    const angle = this.angleFromPositions(cx, cy, ex, ey);
    return angle < 0 ? 360 + angle : angle;
  },
  range(start = 0, stop = null, step = 1) {
    let [_start, _stop] = [0, start];
    if (stop !== null) {
      [_start, _stop] = [start, stop];
    }
    const length = Math.max(Math.ceil((_stop - _start) / step), 0);
    const range = Array(length);

    for (let idx = 0; idx < length; idx++, _start += step) {
      range[idx] = _start;
    }

    return range;
  },
  round(number, decimals) {
    if (!isNaN(parseFloat(number)) && isFinite(number)) {
      const decimalPower = Math.pow(10, decimals);
      return Math.round(parseFloat(number) * decimalPower) / decimalPower;
    }
    return NaN;
  },
  getViewport() {
    return {
      "height": window.innerHeight || document.documentElement.offsetHeight,
      "width": window.innerWidth || document.documentElement.offsetWidth
    };
  },
  cloneObject(object) {
    return JSON.parse(JSON.stringify(object));
  },
  inputTypeForPrototype(prototype) {
    if (prototype === Date) {
      return "date";
    }
    if (prototype === Number) {
      return "number";
    }
    if (prototype === Boolean) {
      return "checkbox";
    }
    return "text";
  },
  prepareValueForInput(value, type) {
    if (type === "date") {
      return new Date(value).toISOString().slice(0, 10);
    }
    if (type === "checkbox") {
      return value ? "on" : null;
    }
    return value;
  },
  getSelection(elem) {
    let start, end, rangeElem, clone;

    if (elem.selectionStart !== undefined) {
      start = elem.selectionStart;
      end = elem.selectionEnd;
    }
    else {
      try {
        elem.focus();
        rangeElem = elem.createTextRange();
        clone = rangeElem.duplicate();

        rangeElem.moveToBookmark(document.selection.createRange().getBookmark());
        clone.setEndPoint('EndToStart', rangeElem);

        start = clone.text.length;
        end = start + rangeElem.text.length;
      }
      catch (e) { /* not focused or not visible */ }
    }

    return { start, end };
  },
  setSelection(elem, selection) {
    let rangeElem;

    try {
      if (elem.selectionStart !== undefined) {
        elem.focus();
        elem.setSelectionRange(selection.start, selection.end);
      }
      else {
        elem.focus();
        rangeElem = elem.createTextRange();
        rangeElem.collapse(true);
        rangeElem.moveStart('character', selection.start);
        rangeElem.moveEnd('character', selection.end - selection.start);
        rangeElem.select();
      }
    }
    catch (e) { /* not focused or not visible */ }
  }
};
