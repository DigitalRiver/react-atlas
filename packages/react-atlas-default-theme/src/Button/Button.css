@import '../styles/variables.css';

/*Main Styles*/
.base {
  composes: button-marg-1  from '../styles.css';
  composes: button-pad-1 from '../styles.css';
  composes: default-text from '../styles.css';
  composes: cursor-pointer from '../styles.css';
  composes: primary-button-border-width from '../styles.css';
  composes: default-font from '../styles.css';
  composes: rounded from '../styles.css';
  text-decoration: none;
  display: inline-block;
  vertical-align: middle;
  line-height: 1.125rem;
  -webkit-appearance: none;
  font-size: 1.0625rem;
  border-style: solid;
}

.uppercase {
  text-transform: uppercase;
}

.base:hover {
  text-decoration: none;
}

.default_btn {
  composes: base;
  composes: hover-bg-gray from '../styles.css';
  composes: hover-white from '../styles.css';
  composes: hover-border-default from '../styles.css';
}
.default_btn:active {
  box-shadow: inset 0 0 0 20rem var(--darken-2),
  inset 0 3px 4px 0 var(--darken-3),
  0 0 1px var(--darken-2);
}
.default_btn:disabled {
  box-shadow: inset 0 0 0 20rem var(--darken-1);
}

.outline {
  composes: base;
  composes: bg-transparent from '../styles.css';
}
.outline:active {
  box-shadow: inset 0 0 0 20rem var(--darken-2),
  inset 0 3px 4px 0 var(--darken-3),
  0 0 1px var(--darken-2);
}
.outline:disabled {
  opacity: .5;
}

/*default button variant*/
.primary {
  composes: bg-primary from '../styles.css';
  composes: hover-bg-primary from '../styles.css';
  composes: border-primary from '../styles.css';
  composes: hover-border-primary from '../styles.css';
  composes: white from '../styles.css';
}
.button {
  composes: base;
}
.secondary {
  composes: bg-secondary from '../styles.css';
  composes: hover-bg-secondary from '../styles.css';
  composes: border-secondary from '../styles.css';
  composes: hover-boder-secondary from '../styles.css';
  composes: inverse-blue-text from '../styles.css';
}

/* TODO: Remove hardcoded color values and
 * compose from styles.css */
.warning {
  composes: base;
  color: #FFFFFF;
  border-color: #F07822;
  background-color: #F07822;
}

/* TODO: Remove hardcoded color values and
 * compose from styles.css */
.warning:hover {
  color: #FFFFFF;
  border-color: #E24024;
  background-color: #E24024;
}

/* TODO: Remove hardcoded color values and
 * compose from styles.css */
.error {
  composes: base;
  color: #FFFFFF;
  border-color: #EA0F43;
  background-color: #EA0F43;
}

/* TODO: Remove hardcoded color values and
 * compose from styles.css */
.error:hover {
  color: #FFFFFF;
  border-color: #BD0A33;
  background-color: #BD0A33;
}

.link {
  composes: base;
  composes: border-none from '../styles.css';
  composes: sky-blue from '../styles.css';
  composes: border-transparent from '../styles.css';
  composes: bg-transparent from '../styles.css';
}
.link:hover{
  text-decoration: underline;
  border-color: transparent;
  background-color: transparent;
  box-shadow: none;
}

/*Outline styles*/
.default_btn_outline {
  composes:outline;
  composes: hover-bg-gray from '../styles.css';
  composes: hover-white from '../styles.css';
  composes: hover-border-default from '../styles.css';
}
.primary_outline{
  composes: outline;
  composes: sky-blue from '../styles.css';
  composes: hover-bg-primary from '../styles.css';
  composes: hover-white from '../styles.css';
  composes: border-primary from '../styles.css';
  composes: hover-border-primary from '../styles.css';
}
.button_outline {
  composes: outline;
  composes: hover-bg-gray from '../styles.css';
  composes: hover-border-default from '../styles.css';
  composes: hover-white from '../styles.css';
}
.secondary_outline {
  composes: outline;
  composes: blue from '../styles.css';
  composes: hover-bg-secondary from '../styles.css';
  composes: border-secondary from '../styles.css';
  composes: hover-white from '../styles.css';
  composes: hover-border-secondary from '../styles.css';
}

.success_outline {
  composes: outline;
  composes: green from '../styles.css';
  composes: hover-bg-green from '../styles.css';
  composes: hover-white from '../styles.css';
  composes: border-green from '../styles.css';
}

.warning_outline {
  composes: outline;
  composes: orange from '../styles.css';
  composes: hover-bg-orange from '../styles.css';
  composes: hover-white from '../styles.css';
  composes: border-orange from '../styles.css';
}

.error_outline {
  composes: outline;
  composes: red from '../styles.css';
  composes: hover-bg-red from '../styles.css';
  composes: hover-white from '../styles.css';
  composes: border-red from '../styles.css';
}

.link_outline {
  composes: outline;
  composes: primary-text from '../styles.css';
  composes: bg-transparent from '../styles.css';
  composes: border-none from '../styles.css';
}
.link_outline:hover {
  text-decoration:underline;
}
.disabled {
  composes: cursor-not-allowed from '../styles.css';
  text-decoration: none;
  opacity: .65;
}

/*button sizes*/
.large {
  padding: 1rem 1.25rem;
  font-size: 1.25rem;
}

.small {
  padding: .25rem .5rem;
  font-size: .85rem;
}

/*block style*/
.block {
  display: block;
  width: 100%;
}

.icon-left {
  margin-right: 1rem;
}