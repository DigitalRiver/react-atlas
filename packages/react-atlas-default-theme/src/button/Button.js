import { Button } from "react-atlas-core";
import style from "./button.css";
import createComponent from "../createComponent";

export default createComponent(Button, style);
