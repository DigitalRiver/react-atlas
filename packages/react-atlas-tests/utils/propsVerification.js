export function verifyPropsDefaultValue(component, expectedProps) {
  //Compares each property value of the component with the expected value

  var result = true;
  var partialResult;

  for (var [compPropName, compPropValue] of expectedProps.entries()) {
    partialResult = component.props()[compPropName] == compPropValue;

    if (!partialResult) {
      console.log(
        "[" +
          compPropName +
          "] -> Got: " +
          component.props()[compPropName] +
          " - expected: " +
          compPropValue
      );
      result = false;
    }
  }
  return result;
}
