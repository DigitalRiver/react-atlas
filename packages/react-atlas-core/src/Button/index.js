import { default as ButtonCore } from "./Button";
export { ButtonCore };
