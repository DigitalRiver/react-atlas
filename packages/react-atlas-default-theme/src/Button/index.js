import * as ButtonStyle from "./Button.css";
export { ButtonStyle };
