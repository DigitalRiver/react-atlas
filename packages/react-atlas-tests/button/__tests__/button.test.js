import React from "react";
import { mount } from "enzyme";
import { default as Button } from "../../../react-atlas-core/src/Button/Button";

import { verifyPropsDefaultValue } from "../../utils/propsVerification";

describe("Test ButtonCore component", () => {
  it("Test default props", function() {
    const result = mount(<Button />);
    expect(result.props().children).toBe("Default Button");
    expect(result.props().outline).toBe(false);
  });

  it("Make sure main style is set correctly.", function() {
    let result = mount(<Button />);
    expect(result.props().primary).toBe(false);

    result = mount(<Button primary />);
    expect(result.props().primary).toBe(true);

    result = mount(<Button secondary />);
    expect(result.props().secondary).toBe(true);

    result = mount(<Button success />);
    expect(result.props().success).toBe(true);

    result = mount(<Button warning />);
    expect(result.props().warning).toBe(true);

    result = mount(<Button danger />);
    expect(result.props().danger).toBe(true);

    result = mount(<Button link />);
    expect(result.props().link).toBe(true);
  });

  describe("Test Button component", () => {
    it("Base case - Primary Button", function() {
      const comp = mount(<Button primary />);
    });

    it("Base case - Secondary Button", function() {
      const comp = mount(<Button secondary />);
    });

    it("Base case - Warning Button", function() {
      const comp = mount(<Button warning />);
    });

    it("Base case - Error Button", function() {
      const comp = mount(<Button error />);
    });

    it("Base case - Link Button", function() {
      const comp = mount(<Button link />);
    });

    it("Base case - Button with icon", function() {
      const comp = mount(<Button icon={"FakeIcon"} />);
    });

    it("Base case - Button with icon 2", function() {
      const comp = mount(
        <Button icon={"FakeIcon"} children={"FakeChildren"} />
      );
    });

    it("Base case - Button default properties", function() {
      const comp = mount(<Button />);
      const expected = new Map([
        ["children", "Default Button"],
        ["outline", false],
        ["primary", false],
        ["secondary", false],
        ["warning", false],
        ["error", false],
        ["link", false],
        ["large", false],
        ["small", false],
        ["disabled", false]
      ]);
      expect(verifyPropsDefaultValue(comp, expected)).toEqual(true);
    });

    it("Base case - Button custom properties", function() {
      const comp = mount(
        <Button secondary={true} warning={true} large={true} disabled={true} />
      );
      const expected = new Map([
        ["children", "Default Button"],
        ["outline", false],
        ["primary", false],
        ["secondary", true],
        ["warning", true],
        ["error", false],
        ["link", false],
        ["large", true],
        ["small", false],
        ["disabled", true]
      ]);
      expect(verifyPropsDefaultValue(comp, expected)).toEqual(true);
    });

    it("Click Button test", function() {
      const comp = mount(
        <Button
          onClick={function() {
            buttonWasClicked = true;
          }}
        />
      );

      let buttonWasClicked = false;

      comp.simulate("click");

      expect(buttonWasClicked).toEqual(true);
    });
  });
});
