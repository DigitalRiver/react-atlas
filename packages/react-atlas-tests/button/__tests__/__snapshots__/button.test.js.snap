// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test correct render Test correct render 1`] = `
<a
  href="http://www.google.com.are"
>
  <button
    className="aClass"
    href="http://www.google.com.are"
    onClick={undefined}
    style="Style"
    styleName="button default_btn"
    type="button"
  >
    Default Button
  </button>
</a>
`;
